dbos/kafka.py

```diff
@@ -0,0 +1,95 @@
+import threading
+import traceback
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Callable, Generator, NoReturn, Optional, Union
+
+from confluent_kafka import Consumer, KafkaError, KafkaException
+from confluent_kafka import Message as CTypeMessage
+
+if TYPE_CHECKING:
+    from dbos.dbos import _DBOSRegistry
+
+from .context import SetWorkflowID
+from .kafka_message import KafkaMessage
+from .logger import dbos_logger
+
+
+def _from_kafka_message(kafka_message: "CTypeMessage") -> KafkaMessage:
+    return KafkaMessage(
+        headers=kafka_message.headers(),
+        key=kafka_message.key(),
+        latency=kafka_message.latency(),
+        leader_epoch=kafka_message.leader_epoch(),
+        offset=kafka_message.offset(),
+        partition=kafka_message.partition(),
+        timestamp=kafka_message.timestamp(),
+        topic=kafka_message.topic(),
+        value=kafka_message.value(),
+    )
+
+
+KafkaConsumerWorkflow = Callable[[KafkaMessage], None]
+
+
+def _kafka_consumer_loop(
+    func: KafkaConsumerWorkflow,
+    config: dict[str, Any],
+    topics: list[str],
+    stop_event: threading.Event,
+) -> None:
+
+    def on_error(err: KafkaError) -> NoReturn:
+        raise KafkaException(err)
+
+    config["error_cb"] = on_error
+    if "auto.offset.reset" not in config:
+        config["auto.offset.reset"] = "earliest"
+
+    consumer = Consumer(config)
+    try:
+        consumer.subscribe(topics)
+        while not stop_event.is_set():
+            cmsg = consumer.poll(1.0)
+
+            if stop_event.is_set():
+                return
+
+            if cmsg is None:
+                continue
+
+            msg_error = cmsg.error()
+            if msg_error is not None:
+                if msg_error.code() == KafkaError._PARTITION_EOF:
```

Comment:
What are the semantics of these errors? Are they something we should send to the consumer or something that can't be handled?

Reply:
I replicated the logic from https://docs.confluent.io/kafka-clients/python/current/overview.html#basic-poll-loop

Reply:
Got it!  But because this is in a thread, we need to log the exception before raising it. We should then decide if we want to kill the consumer forever (if the error is nonrecoverable, this is fine as long as it's logged) or continue past the exception.